### PR TITLE
Add Block Transform to transform Embed blocks into Paragraph blocks.

### LINKED ISSUE #15102: Option to use text link instead of auto embed.
**Is your feature request related to a problem? Please describe.**
Adding links. If an link goes to another WordPress site we will see an auto embed such as this:

Screen Shot 2019-04-22 at 11 12 35

A mix of auto embeds and text links.

Screen Shot 2019-04-22 at 11 22 24

**Describe the solution you'd like**
I would like the option to include an auto embed OR a text link. As some links above are automatically embedded showing an article preview while other links just show a text link. 

**Describe alternatives you've considered**
I added the following code from my tutorial here:
https://easywebdesigntutorials.com/adjusting-the-auto-embeds-of-wordpress-site-links-on-your-site/

By adding code from the above tutorial it forces the link dialog box to open for the links that were auto embedded. I can then select to convert the auto embeds to regular text links. One should not need to add the code to turn off auto embeds.
Screen Shot 2019-04-22 at 11 39 36

### PR DESCRIPTION
**Pull request by desaiuditd**
## Description
Fixes #15102 by providing a block transform to transform Embeds into Paragraphs.

Ref: https://github.com/WordPress/gutenberg/issues/15102#issuecomment-485428925

## How has this been tested?

Pasting a URL into the Paragraph block automatically converts into an Embed blocks. After that, now, I'm able to transform the same Embed block into a Paragraph block. Performing an Undo action brings the Embed block back.

While working on this, I found a bug that none of the transformations of `core/embed` block were applied to `core-embed/common/*` and `core-embed/others/*` blocks. So I've fixed that as well.

## Types of changes

While converting the Embed into Paragraph, if the caption is present, then I'm using that as the text for the link in the Paragraph. That seemed like a sensible behaviour without the data loss for the user. Happy to defer on someone who's more close to this.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-library/src/embed/core-embeds.js [hunk 1]
```diff
@@ -151,15 +151,17 @@ export const others = [
...
+				from: [ {
+					type: 'block',
+					blocks: [ 'core-embed/polldaddy' ],
+					transform: ( content ) => {
+						return createBlock( 'core-embed/crowdsignal', {
+							content,
+						} );
+					},
+				} ],
+			},
```
Comment (desaiuditd) [inline packages/block-library/src/embed/core-embeds.js]: Not sure if this is a breaking change or not. Simply changed the schema of JSON to make it consistent with other transforms.

File: packages/block-library/src/embed/index.js [hunk 2]
```diff
@@ -19,35 +24,37 @@ export const settings = getEmbedBlockSettings( {
...
+		const embedSettings = getEmbedBlockSettings( embedDefinition.settings );
 		return {
 			...embedDefinition,
-			settings: getEmbedBlockSettings( embedDefinition.settings ),
+			settings: {
+				...embedSettings,
+				transforms: {
+					from: concat( get( transforms, 'from', [] ), get( embedSettings, 'transforms.from', [] ) ).filter( Boolean ),
+					to: concat( get( transforms, 'to', [] ), get( embedSettings, 'transforms.to', [] ) ).filter( Boolean ),
+				},
```
Comment (desaiuditd) [inline packages/block-library/src/embed/index.js]: Basically, apply combined transformations of `core/embed` as well as other `core-embed/common/*`

**Comment by paaljoachim** [comment]
Hey @desaiuditd 

Here is even more information: https://github.com/WordPress/gutenberg/issues/21029

**Comment by desaiuditd** [comment]
This is good for review again.

**Comment by paaljoachim** [comment]
Hey Udit.

I tested the PR here:
http://gutenberg.run/17413

I then pasted the following links:
https://make.wordpress.org/
https://make.wordpress.org/core/
https://make.wordpress.org/test/2020/03/11/gutenberg-usability-testing-for-january-2020/

All three links brought up a following type of Embed placeholder box:
Screen Shot 2020-04-05 at 11 55 55

I clicked to embed but none of them could be embedded.
I clicked to preview the frontend and noticed that all are text links.


--------------------

Here is my suggestion on how to simplify the process for a user adding a link.

The user pastes inn a link and it automatically shows up as a text link.
As all links can show up as text links. 

But some links that can also be embedded. For these special links there can be an option to transform the text link into an embed through for instance the Transform area.

The transform area.
Screen Shot 2020-04-05 at 12 12 53

Transforming a text link into an embed where that can happen.

I will ping Miguel, so we can discuss it further here.
@mcsf

**Comment by desaiuditd** [comment]
> All three links brought up a following type of Embed placeholder box:

I tried this on preview server as well. But I also noticed these oembed api request failures.

image

That could be the reason why Gutenberg failed to convert them into embeds on paste.

**Comment by desaiuditd** [comment]
> The user pastes inn a link and it automatically shows up as a text link.
> As all links can show up as text links.
> 
> But some links that can also be embedded. For these special links there can be an option to transform the text link into an embed through for instance the Transform area.

This changes the current default behaviour, which @mcsf mentioned.

> It always tries to embed, but this process may fail depending on the source site.

So can we let the current behaviour be as is? And when the process may fail, user has two options to revert. Either transform to paragraph block or click `Convert to link` button.

**Comment by desaiuditd** [comment]
I've one question for Option 1 that @paaljoachim has suggested.

> The user pastes inn a link and it automatically shows up as a text link.
As all links can show up as text links.
>
> But some links that can also be embedded. For these special links there can be an option to transform the text link into an embed through for instance the Transform area.

how can I check which embed transformation to allow? E.g., if I paste a WordPress Make URL, it’s not probably right to show Youtube embed in the transformation popup, right?

Check this Gif. I got this far. But not sure how to enable only valid embeds out of all. It’s even erroring out for any mismatched embed.



@mcsf Do you have any thoughts on this?

Meanwhile, I will try and see if I can get Option 2 working.

**Comment by desaiuditd** [comment]
@paaljoachim @mcsf This is good for review again. I was able to achieve Option 1 suggested by Paal.

In order to do that, I've removed `__unstableEmbedURLOnPaste`. @ellatrix It would be good if you review this as well. Thanks.

**Review by mcsf — CHANGES_REQUESTED** [review]
Hi, thanks for helping out here.

This PR is changing a lot of core behaviours that the discussion in the parent issue didn't create consent for.

In particular, the default behaviour of pasting URLs as embeds should be preserved. This isn't just an established (and expected by most) behaviour of the block editor, it is also a reflection of what the classic editor has done for a long time.

Also very importantly, the WordPress back end itself treats URLs on their own line as embeds. By disabling the automatic pasting of URLs as embeds in the block editor, we don't actually prevent embeds from appearing on the rendered page — thus the problem isn't solved, and indeed is made worse by breaking the user's trust that the block editor accurately represents the page.

Things to revert:

* The removal of the logic around `__unstableEmbedURLOnPaste` in Paragraph and RichText
* The removal of iframe-related integration tests
* The refactoring around embeds. The embeds code base could use an important refresh, but until then it's better to leave it as is, and avoid abstractions like `_mergeTransforms`.

**The core feature to expect from this PR is the ability to convert an embed into a block of text that links to the URL in question.**

Thanks for addressing this, and I apologise for not being able to review this later.

**Comment by paaljoachim** [comment]
Hi Miguel @mcsf 

That sounds like option 2 which I mentioned above.
Option 2.
Pasting a link when it is not embeddable automatically adds it as a text link.
Pasting a link when it is embeddable opens a placeholder box to where the user can choose to embed or choose to paste it as a text link. (There also needs to be a simple way to revert to either text or embed link.)

**Comment by desaiuditd** [comment]
@paaljoachim if we do this

> Pasting a link when it is not embeddable automatically adds it as a text link.

then, why there's a need to do this?

> Pasting a link when it is embeddable opens a placeholder box to where the user can choose to embed or choose to paste it as a text link.

We are already doing below by providing block transform to paragraph. So they can easily go back to text link.

> There also needs to be a simple way to revert to either text or embed link.

As Miguel said, the default behaviour of pasting URLs as embeds should be preserved. Which is automatically converting a link into an embed.

**Comment by desaiuditd** [comment]
> it's better to leave it as is, and avoid abstractions like `_mergeTransforms`.

@mcsf I was trying to follow DRY because it was same logic for `commonEmbeds` and `otherEmbeds` embeds. Also, couple of embeds like Speaker Deck and Crowdsignal were coming with their own transforms. So they needed to be merged with the default transforms. Hence, `_mergeTransforms`, to retain embed specific transforms as well as the default one.

> **The core feature to expect from this PR is the ability to convert an embed into a block of text that links to the URL in question.**

@mcsf Do you think, I should raise another PR to fix #15102 ?

**Comment by paaljoachim** [comment]
I think I am just confusing myself here..... I might be running around in circles... sorry about that.
Bottom line is the initial post here: https://github.com/WordPress/gutenberg/issues/21029
and @mapk 's response.

I am happy to see that @mcsf Miguel is giving some feedback so we get the concept/development on the right course. Thank you Miguel.

**Comment by paaljoachim** [comment]
So the process should be: 
Link -> Can it embed? -> If yes -> Add a placeholder asking if the user wants to embed or convert to text link.
Link -> Can it embed? -> If no -> Automatically add a text link.

@mcsf @desaiuditd @mapk

**Comment by desaiuditd** [comment]
> So the process should be:
> Link -> Can it embed? -> If yes -> Add a placeholder asking if the user wants to embed or convert to text link.
> Link -> Can it embed? -> If no -> Automatically add a text link.
> 
> @mcsf @desaiuditd @mapk

Cool. Let me raise another PR to fix this (#21029), so we can have some separation of concern. This PR can fix  #15102 in isolation.

**Comment by paaljoachim** [comment]
I wrote the following in the core editor slack channel a few hours ago and will add it here as well.
So from my understanding of this PR. It takes a part various embeds. Then @desaiuditd will create a follow up PR where embeds are put together even better again. We are looking at all pasted links that embed such as a video etc to work as normal. But pasting a WP post content which normally just adds an embed will instead give the user a placeholder option to either embed the link or to show it as a text link. If someone can check out the PR and let us know if this is the correct approach that would be great. Thanks for any feedback and thoughts shared.

**Comment by mcsf** [comment]
@desaiuditd:

> @mcsf I was trying to follow DRY because it was same logic for `commonEmbeds` and `otherEmbeds` embeds. Also, couple of embeds like Speaker Deck and Crowdsignal were coming with their own transforms. So they needed to be merged with the default transforms. Hence, `_mergeTransforms`, to retain embed specific transforms as well as the default one.

Right, the custom `transforms` for Speaker Deck and Crowdsignal are broken, but are also unnecessary, since — in my testing — the editor corrects the embed type based on the provided patterns anyway. I think we should just remove them, but that belongs in a separate PR.

I appreciate the concern for DRY, although it's a constant tug between DRY and the visual and immediate clarity that sometimes an inline transformation can best provide.

In this case, we don't need to touch any of it, so I recommend reverting.

> > **The core feature to expect from this PR is the ability to convert an embed into a block of text that links to the URL in question.**
> 
> @mcsf Do you think, I should raise another PR to add smart handling of Embeds as per what Paal is suggesting? That way, this PR can just be about adding new block transform which converts embed into a paragraph with link to the embed url.

Yes, it's best to keep things focused. I personally am not convinced about smart handling of embeds, at least in the way it was it presented, but it can be put up for discussion.

Thanks for coming back to this PR, for making these changes and for keeping things organised!

**Comment by desaiuditd** [comment]
@mcsf This is good for review again. I have made the requested changes.

**Review by mcsf — APPROVED** [review]
Thanks for this fix!

**Comment by paaljoachim** [comment]
**NB!** 
This PR should not be included into 8.1, as the PR tears a part embeds and we need the PR that rebuilds embeds to be merged at the same time, so that we do not break Youtube and other embeds.

**Comment by mcsf** [comment]
I'm not experiencing any regressions so far. I've compared 81d4f4c22f (PR head) with the base branch at 601ee4b70d, and I've tested again with the commit to master at 48aba1ea96e9209b36c0fc55df5969e87e54ed2d.

You may be experiencing this for being stuck with a stale build in gutenberg.run, something which should be fixed soon ([Slack link](https://wordpress.slack.com/archives/C02QB2JS7/p1589200085030500)).

**Comment by paaljoachim** [comment]
Retesting again in gutenberg.run.
I just retested again (in Chrome), and I am getting the same result.

I will retest in another browser (Firefox). Same result.

Screen Shot 2020-05-11 at 15 52 28

**Comment by desaiuditd** [comment]
@paaljoachim This PR was never meant to fix the smart auto embedding which you suggested earlier. This PR only allows user to convert an existing embed block into a paragraph block. So it only fixes #15102 in isolation.

The approach you suggested is still being discussed at #21029.

**Comment by paaljoachim** [comment]
Hey @desaiuditd 

My concern during the test at gutenberg.run was that one is not able to embed even the Youtube embeds that I tested with. 

So from what you are saying (hopefully) the above would be an error.
One should be able to embed. Then take any embed and convert it into a paragraph block.
Is this correct?

File: packages/block-library/src/embed/transforms.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		},
+	],
+	to: [
+		{
+			type: 'block',
+			blocks: [ 'core/paragraph' ],
+			transform: ( { url, caption } ) => {
+				const link = <a href={ url }>{ caption || url }</a>;
+				return createBlock( 'core/paragraph', {
+					content: renderToString( link ),
```
Comment (ellatrix) [inline packages/block-library/src/embed/transforms.js:31]: Why not write the html as a string?

**Comment by jorgefilipecosta** [comment]
The PR seems to work well in my tests. I am also able to reproduce the issue @paaljoachim referred but only on gutenberg.run. @aduth verified gutenberg.run has an issue in requests to the endpoint `/wp-json/oembed/1.0/proxy?url=`.

File: packages/block-library/src/embed/transforms.js (hunk 3 above)
Comment (mcsf) [reply packages/block-library/src/embed/transforms.js:31]: I don't mind the added peace of mind of running this through the element serialiser. Do you have drawbacks in mind?

**Comment by mcsf** [comment]
Thanks for debugging that! Seems we can move on with the release.

**Comment by paaljoachim** [comment]
Testing with 8.1.0-rc.1

Using URL's that automatically create an embed. Such as
https://make.wordpress.org/core/2020/05/12/editor-chat-agenda-13-may-2020/

The embed:
Screen Shot 2020-05-12 at 16 07 44

Transform to Paragraph Block:
Screen Shot 2020-05-12 at 16 07 51

Transformed Embed block which has not turned into a Paragraph Block:
Screen Shot 2020-05-12 at 16 07 56

**Pull request merged**